akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala

```diff
@@ -546,6 +546,7 @@ private[persistence] trait Eventsourced extends Snapshotter with Stash with Stas
           updateLastSequenceNr(p)
           try {
             peekApplyHandler(p.payload)
+            onPersistSuccess(p.payload)
```

Comment:
No need to be sorry, great thought. I totally overlooked that one. Currently, when the callback throws, all messages get unstashed and the exception is thrown so that the Actor will (if not implemented a different supervisor strategy) restart.

What should really happen? The persist callback failed (unable to update the internal state for instance). What does this mean for the `onPersistSuccess` if it uses the internal state? 
In my case I want to publish on the `EventStream` that a message is successfully persisted, so the hook should be called always when persistence succeeded, which is the case here. 

In my current user code and in your proposed code (`(publishEvent andThen somethingElse)`) the hook is called first. This essentially turns the problem around, but might be more straight forward to reason about. I've got the feeling that this might be the best solution. Always call the `onPersistSuccess` first and when that succeeded call the persist callback. Then document that the `onPersistSuccess` is called before the callback. When the `onPersistSuccess` throws an Exception the callback will not be called. 

I don't think it is possible to make this foolproof, but this seems to be the most safe way and straight forward solution to me. What do you think?